### ISSUE TITLE
TF Keras inference is way slower than Numpy

### ISSUE THREAD

**Issue by alexschoeberl** [tensorflow/tensorflow#17207]
I'm working on a reinforcement learning model implemented with Keras and Tensorflow. I have to do frequent calls to model.predict() on single inputs.

While testing inference on a simple pretrained model, I noticed that using Keras' model.predict is WAY slower than just using Numpy on stored weights. Why is it that slow and how can I accelerate it? Using pure Numpy is not viable for complex models.

    import timeit
    import numpy as np
    from tensorflow.python.keras.models import Sequential
    from tensorflow.python.keras.layers import Dense
    
    w = np.array([[-1., 1., 0., 0.], [0., 0., -1., 1.]]).T
    b = np.array([ 15., -15., -21., 21.])
    
    model = Sequential()
    model.add(Dense(4, input_dim=2, activation='linear'))
    model.layers[0].set_weights([w.T, b])
    model.compile(loss='mse', optimizer='adam')
    
    state = np.array([-23.5, 17.8])
    
    def predict_very_slow():
        return model.predict(state[np.newaxis])[0]
    
    def predict_slow():
        ws = model.layers[0].get_weights()
        return np.matmul(ws[0].T, state) + ws[1]
    
    def predict_fast():
        return np.matmul(w, state) + b
    
    print(
        timeit.timeit(predict_very_slow, number=10000),
        timeit.timeit(predict_slow, number=10000),
        timeit.timeit(predict_fast, number=10000)
    )
    # 5.168972805004538 1.6963867129435828 0.021918574168087623
    # 5.461319456664639 1.5491559107269515 0.021502970783442876

I'm using Tensorflow for CPU, version 1.5.0 installed from pypi for python 3.5 on Windows 10.

**Comment by bignamehyp** [comment]
If it's just matmul, why not use tf.matmul?


import timeit
import numpy as np
import tensorflow as tf

w = np.array([[-1., 1., 0., 0.], [0., 0., -1., 1.]]).T
b = np.array([[ 15., -15., -21., 21.]])
s = np.array([[-23.5, 17.8]])
u = tf.matmul(w, s, transpose_b=True) + b.T
sess = tf.Session()

rand_array = np.random.rand(size, size)

start_time = time.time()
for _ in xrange(10):
  np.dot(np.dot(rand_array,rand_array), rand_array)

print("--- %s seconds numpy multiply ---" % (time.time() - start_time))

start_time = time.time()
for _ in xrange(10):
  sess.run(z, feed_dict={x: rand_array})

print("--- %s seconds tensorflow---" % (time.time() - start_time))

This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by alexschoeberl** [comment]
It's just one matmul in this demonstration, in reality more complex networks are used.
You can only feed one input/state at a time, because the next state depends on the result.
I already asked this question on [StackOverflow](https://stackoverflow.com/questions/48796619/why-is-tf-keras-inference-way-slower-than-numpy-operations).
